crates/sui-core/src/authority_active/checkpoint_executor/mod.rs

```diff
@@ -0,0 +1,295 @@
+//! for catching up to the current checkpoint sequence number of the network
+//! as quickly as possible so that a newly joined, or recovering Node can
+//! participate in a timely manner. To that end, CheckpointExecutor attempts
+//! to saturate the CPU with executor tasks (one per checkpoint), each of which
+//! handle scheduling and awaiting checkpoint transaction execution.
+//!
+//! CheckpointExecutor is made recoverable in the event of Node shutdown by way of a watermark,
+//! highest_executed_checkpoint, which is guaranteed to be updated sequentially in order,
+//! despite checkpoints themselves potentially being executed nonsequentially and in parallel.
+//! CheckpointExecutor parallelizes checkpoints of the same epoch as much as possible, and
+//! handles epoch boundaries by calling to reconfig once all checkpoints of an epoch have finished
+//! executing.
+
+use core::panic;
+use std::{sync::Arc, time::Duration};
+
+use futures::stream::FuturesOrdered;
+use mysten_metrics::spawn_monitored_task;
+use sui_types::{
+    base_types::{ExecutionDigests, TransactionDigest},
+    crypto::AuthorityPublicKeyBytes,
+    error::{SuiError, SuiResult},
+    messages::TransactionEffects,
+    messages_checkpoint::{CheckpointSequenceNumber, VerifiedCheckpoint},
+};
+use tokio::{
+    sync::broadcast::{self, error::RecvError},
+    task::JoinHandle,
+};
+use tokio_stream::StreamExt;
+use tracing::{error, info, warn};
+use typed_store::rocks::TypedStoreError;
+
+use crate::{
+    authority::{AuthorityState, EffectsNotifyRead},
+    checkpoints::{CheckpointMetrics, CheckpointStore},
+};
+
+#[cfg(test)]
+pub(crate) mod tests;
+
+const TASKS_PER_CORE: usize = 1;
+
+type CheckpointExecutionBuffer = FuturesOrdered<
+    JoinHandle<(
+        u64,
+        VerifiedCheckpoint,
+        Option<Vec<(AuthorityPublicKeyBytes, u64)>>,
+    )>,
+>;
+
+pub struct CheckpointExecutor {
+    mailbox: broadcast::Receiver<VerifiedCheckpoint>,
+    checkpoint_store: Arc<CheckpointStore>,
+    authority_state: Arc<AuthorityState>,
+    highest_scheduled_checkpoint: Option<CheckpointSequenceNumber>,
+    end_of_epoch: bool,
+    metrics: Arc<CheckpointMetrics>,
+}
+
+impl CheckpointExecutor {
+    pub fn new(
+        mailbox: broadcast::Receiver<VerifiedCheckpoint>,
+        checkpoint_store: Arc<CheckpointStore>,
+        authority_state: Arc<AuthorityState>,
+        metrics: Arc<CheckpointMetrics>,
+    ) -> Result<Self, TypedStoreError> {
+        Ok(Self {
+            mailbox,
+            checkpoint_store,
+            authority_state,
+            highest_scheduled_checkpoint: None,
+            end_of_epoch: false,
+            metrics,
+        })
+    }
+
+    pub async fn run(mut self) {
+        while let Some((last_checkpoint, next_committee)) =
+            self.execute_checkpoints_for_epoch().await
+        {
+            reconfig(next_committee).await;
+            self.checkpoint_store
```

Comment:
So we update_highest_executed_checkpoint after reconfig. Why is that the case?
Imagine the follow scenario: we executed the last checkpoint, finished reconfiguring the state and then something bad happened and the machine shuts down. We restart in a state where we are in a new epoch but the last checkpoint from the previous epoch is still marked as not-executed. Would that be an issue?

Reply:
the alternative is also bad, we mark the last checkpoint in an epoch executed, then we reconfig but fail to do so. Then when we startup we try to execute the first checkpoint in the next epoch but reconfiguration hasn't happened yet fully. This is a nuanced problem and one that i'm sure we'll want to test thoroughly :D

Reply:
So perhaps we have two separate marks, one that marks the last executed checkpoint, and one that marks reconfiguration success?

Reply:
That is an easily recoverable state. We simply pick up on the same checkpoint, execute, and redo reconfig (which will essentially be a no-op). This is better than the alternative scenario that could arise if we update the watermark after exec but before reconfig -- we execute, update watermark, then node shuts down. In this case, we are in the previous epoch and need to reconfig, but we don't know it (since we only have the committee for next epoch from the last checkpoint of epoch). In this case, we would carry on executing subsequent checkpoints with invalid configuration.

This is the impetus of using the watermark as a finality

Reply:
But I would assume that you would no longer be able to successfully schedule the checkpoint since its epoch no longer matches with the state epoch

Reply:
Good point. I think the 2 watermarks idea is a good one. Will do

Reply:
Rather than adding a second watermark, I decided to handle this case simply by checking for it at startup and immediately doing reconfig if needed.